### PR TITLE
DD-667 transfer to vault m2 extract metadata update db

### PR DESCRIPTION
**Pull request by mderuijter**
Configure `config.yml`:
```
inboxes:
  - name: DvInstance1
    path: src/test/resources/data/DvInstance1
#  - name: DS_Archaeology
#    path: /path/to/inbox/archaeology
#  - name: DS_SSH
#    path: /path/to/inbox/ssh
```
run `mvnci`
run `start-hsqldb.sh` (might need to be reset to avoid SQL constraint exceptions)
run `start-service.sh`

### CODE REVIEW

File: src/main/java/nl/knaw/dans/ttv/core/TransferTask.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        extractMetadata();
+        return transferItem;
+    }
+
+    @UnitOfWork
+    public void extractMetadata() throws IOException {
+        ZipFile datasetVersionExport = new ZipFile(Paths.get(transferItem.getDveFilePath()).toFile());
+        String metadataFilePath = Objects.requireNonNull(datasetVersionExport.stream()
+                .filter(zipEntry -> zipEntry.getName().endsWith(".jsonld"))
+                .findAny().orElse(null)).getName();
```
Comment (janvanmansum) [inline src/main/java/nl/knaw/dans/ttv/core/TransferTask.java:52]: So, if the `.jsonld` file is not found a NPE is thrown? Can you make sure that something more descriptive is logged?

File: src/main/java/nl/knaw/dans/ttv/core/TransferTask.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                .filter(zipEntry -> zipEntry.getName().contains("pid-mapping.txt"))
+                .findAny().orElse(null)).getName();
+
+        byte[] pidMapping = IOUtils.toByteArray(datasetVersionExport.getInputStream(datasetVersionExport.getEntry(pidMappingFilePath)));
+        byte[] oaiOre = IOUtils.toByteArray(datasetVersionExport.getInputStream(datasetVersionExport.getEntry(metadataFilePath)));
+
+        ObjectMapper objectMapper = new ObjectMapper();
+        JsonNode jsonNode = objectMapper.readTree(oaiOre);
+
+        String nbn = jsonNode.get("ore:describes").get("dansDataVaultMetadata:NBN").toString();
```
Comment (janvanmansum) [inline src/main/java/nl/knaw/dans/ttv/core/TransferTask.java:63]: A lot of potential NPE's here.

File: src/main/java/nl/knaw/dans/ttv/core/TransferTask.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    @Override
+    public TransferItem call() throws Exception {
+        log.info("Running task" + this);
+        extractMetadata();
+        return transferItem;
+    }
+
+    @UnitOfWork
```
Comment (janvanmansum) [inline src/main/java/nl/knaw/dans/ttv/core/TransferTask.java:47]: You could limit the scope of the unit of work annotation. Not sure if it is a big deal, just saying ..

File: src/main/java/nl/knaw/dans/ttv/core/TransferTask.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                .filter(zipEntry -> zipEntry.getName().endsWith(".jsonld"))
+                .findAny().orElse(null)).getName();
+        String pidMappingFilePath = Objects.requireNonNull(datasetVersionExport.stream()
+                .filter(zipEntry -> zipEntry.getName().contains("pid-mapping.txt"))
+                .findAny().orElse(null)).getName();
+
+        byte[] pidMapping = IOUtils.toByteArray(datasetVersionExport.getInputStream(datasetVersionExport.getEntry(pidMappingFilePath)));
+        byte[] oaiOre = IOUtils.toByteArray(datasetVersionExport.getInputStream(datasetVersionExport.getEntry(metadataFilePath)));
+
+        ObjectMapper objectMapper = new ObjectMapper();
```
Comment (janvanmansum) [inline src/main/java/nl/knaw/dans/ttv/core/TransferTask.java:60]: Can we reuse this ObjectMapper? Seems a bit wastewful to create it on every call. Also, DropWizard creates one for the application anyway. Can we somehow use that one?

**Review by janvanmansum — CHANGES_REQUESTED** [review]
Some comments, see inline. Also, please merge back master, so the diff will be smaller.

**Comment by mderuijter** [comment]
> Some comments, see inline. Also, please merge back master, so the diff will be smaller.

@janvanmansum It's weird, merged back master but the diffs are still there...

File: src/main/java/nl/knaw/dans/ttv/core/TransferTask.java (hunk 4 above)
Comment (mderuijter) [reply src/main/java/nl/knaw/dans/ttv/core/TransferTask.java:59]: I'll look into it

File: src/main/java/nl/knaw/dans/ttv/core/TransferTask.java (hunk 3 above)
Comment (mderuijter) [reply src/main/java/nl/knaw/dans/ttv/core/TransferTask.java:46]: @janvanmansum what do you mean? It's limited to this method? Can I limit it even more?

**Pull request merged**